Issue1492 reduced order iso13790

amaccarini opened the pull request:
This package contains reduced-order building energy models based on the ISO 13790 standard.

KremerMartin commented on IBPSA/ThermalZones/ReducedOrder/RC/OneElement.mo:
```diff
@@ -386,7 +386,7 @@ equation
     annotation (Line(points={{-20,-16},{-20,0},{80,0}}, color={191,0,0}));
   connect(senTAir.T, TAir)
-    annotation (Line(points={{100,0},{108,0},{108,160},{250,160}},
+    annotation (Line(points={{101,0},{108,0},{108,160},{250,160}},
```
I think this little changes shouldn't be merged, right?

amaccarini replied:
No, this actually belongs to another package. Not sure how it happened.

KremerMartin commented on IBPSA/package.mo:
```diff
@@ -486,7 +486,7 @@ annotation (
 versionDate="2018-09-28",
 dateModified = "2018-09-28",
-uses(Modelica(version="4.0.0")),
+uses(Modelica(version="4.0.0"), Buildings(version="8.0.0")),
```
I think the _Buildings_-Library should not be used by the _IBPSA_. Normally it is vice versa. I cannot find any model implemented by this PR that uses the _Buildings_.

amaccarini replied:
Not sure what to do here. Should I just remove that String from the code?

KremerMartin replied:
I would suggest to try and remove "Buildings(version="8.0.0")" and check the IBPSA package. If no error occurs, the library isn't used in any model of the IBPSA and all is fine without the additional string. :)

KremerMartin reviewed (changes requested):
Thanks, @amaccarini for your great work here. All in all it seems to be quite a good model. I have still some changes, that you should address. Mostly it is linked to the StyleGuide, so small changes.

- [x] Please rework your models to fulfil the requirements of the StyleGuide
- [ ] Add documentation to the instances, variables and parameters.
- [ ] Check, if it isn't more applicable to summarize all building data (heat capacity, u-values, areas) in the data records or check if there is really need for a record only containing one parameter.
- [ ] Check, if the examples and validation can be simplified (maybe 'multiplex'-model is not needed)

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GainMass.mo:
```diff
@@ -0,0 +1,50 @@
+within IBPSA.ThermalZones.ISO13790.BaseClasses;
+model GainMass "Mass node heat flow"
+
+  Real Am;
```
I added documentation for Am

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GainMass.mo:
```diff
@@ -0,0 +1,50 @@
+within IBPSA.ThermalZones.ISO13790.BaseClasses;
+model GainMass "Mass node heat flow"
+
+  Real Am;
```
I changed the name from Am to AMas

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GainMass.mo:
```diff
@@ -0,0 +1,50 @@
+  Real Am;
+  parameter Real At;
+  parameter Real HwinG;
+  parameter Real f_ms;
```
Changed f_ms to facMas

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GainMass.mo:
```diff
@@ -0,0 +1,50 @@
+    annotation (Placement(transformation(extent={{-140,-60},{-100,-20}})));
+  Modelica.Blocks.Interfaces.RealInput solG
+    annotation (Placement(transformation(extent={{-140,-100},{-100,-60}})));
+  Modelica.Blocks.Interfaces.RealOutput y=((Am/At)*(0.5*intG + solG))
```
At (which now has become ATot) cannot be zero, as it is not a direct user input, but it is calculated as the sum of all surfaces facing the thermal zone. This can only be zero if all surfaces are zero, which basically cannot happen.

KremerMartin replied:
That's right. It does not make sense to define a thermal zone without any surface areas. Maybe it would be useful to add an assertion for this case. In that way the user gets a more informative error message, if he does that mistake. (e.g. "At least one surface needs an area greater than zero").

mwetter replied:
The multiplex in models such as `IBPSA.ThermalZones.ISO13790.Examples.HeatingCooling` is not needed, but it does not hurt. Either way is fine.

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GainSurface.mo:
```diff
@@ -0,0 +1,51 @@
+within IBPSA.ThermalZones.ISO13790.BaseClasses;
+model GainSurface "Surface node heat flow"
+
+  Real Am;
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GainSurface.mo:
```diff
@@ -0,0 +1,51 @@
+    annotation (Placement(transformation(extent={{-140,-60},{-100,-20}})));
+  Modelica.Blocks.Interfaces.RealInput solG
+    annotation (Placement(transformation(extent={{-140,-100},{-100,-60}})));
+  Modelica.Blocks.Interfaces.RealOutput y=(1 - Am/At - HwinG/(9.1*At))*(
```
Made a comment on this before

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GlazedElements.mo:
```diff
@@ -0,0 +1,301 @@
+  IBPSA.BoundaryConditions.WeatherData.Bus weaBus annotation (Placement(
+        transformation(extent={{-150,-12},{-110,28}}),iconTransformation(extent={{-150,
+            -10},{-130,10}})));
+  IBPSA.BoundaryConditions.SolarIrradiation.DirectTiltedSurface     HDirTil(
```
Not sure I understand what you mean.

KremerMartin replied:
I see your point. Adhoc I do not have a good suggestion for solving that.

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GlazedElements.mo:
```diff
@@ -0,0 +1,301 @@
+within IBPSA.ThermalZones.ISO13790.BaseClasses;
+model GlazedElements "Solar heat gains of glazed elements"
+  IBPSA.BoundaryConditions.WeatherData.Bus weaBus annotation (Placement(
```
Added documentation to Bus

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GlazedElements.mo:
```diff
@@ -0,0 +1,301 @@
+    azi=surfaceAzimuth[4])
+    annotation (Placement(transformation(extent={{-90,-100},{-70,-80}})));
+
+  parameter Real groundReflectance = 0.2 "Ground reflectance"
```
Changed

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GlazedElements.mo:
```diff
@@ -0,0 +1,301 @@
+
+  parameter Real groundReflectance = 0.2 "Ground reflectance"
+      annotation(Evaluate=true, Dialog(tab = "General", group = "Location"));
+  parameter Modelica.Units.SI.Area[4] Awin={0,0,6,0} "Areas of windows"
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GlazedElements.mo:
```diff
@@ -0,0 +1,301 @@
+      annotation(Evaluate=true, Dialog(tab = "General", group = "Location"));
+  parameter Modelica.Units.SI.Area[4] Awin={0,0,6,0} "Areas of windows"
+    annotation (Evaluate=true, Dialog(tab="General", group="Window data"));
+  parameter Real winFrame = 0.01 "Frame fraction of windows"
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GlazedElements.mo:
```diff
@@ -0,0 +1,301 @@
+    annotation (Placement(transformation(extent={{72,-90},{92,-70}})));
+  Modelica.Blocks.Math.Sum sum(nin=4)
+    annotation (Placement(transformation(extent={{104,-10},{124,10}})));
+  Modelica.Blocks.Interfaces.RealOutput y
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GlazedElements.mo:
```diff
@@ -0,0 +1,301 @@
+Mass data for heavy building
+</p>
+</html>"));
+end GlazedElements;
```
Good point. The thing is that at the moment I am not sure which variables could be useful for users. I probably need to use the models a bit more in order to figure out this aspect. I will keep this in mind for a future revision of the models. Do you have some specific examples of variables that could be protected?

KremerMartin replied:
In this very model, I would for example declare the `sum`-model or `fraWinWest` as protected since it only sums up already given/reachable variables. Another example are constants (not in this very model).
But of course this could be optimised by using the models.

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/OpaqueElements.mo:
```diff
@@ -0,0 +1,507 @@
+  IBPSA.BoundaryConditions.WeatherData.Bus weaBus annotation (Placement(
+        transformation(extent={{-200,-20},{-160,20}}),iconTransformation(extent={{-230,
+            -40},{-210,-20}})));
+  IBPSA.BoundaryConditions.SolarIrradiation.DirectTiltedSurface HDirTil(
```
Left a comment on this aspect before

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/OpaqueElements.mo:
```diff
@@ -0,0 +1,507 @@
+    azi=surfaceAzimuth[4])
+    annotation (Placement(transformation(extent={{-120,-100},{-100,-80}})));
+
+  parameter Real groundReflectance = 0.2 "Ground reflectance"
```
Fixed

amaccarini commented on IBPSA/ThermalZones/ISO13790/Data/BESTEST600.mo:
```diff
@@ -0,0 +1,17 @@
+within IBPSA.ThermalZones.ISO13790.Data;
```
The thing is that I think it is useful to add this record as it directly connect the wording "light", "medium", heavy" to a value of heat capacity. Such values are typically not known by users, so it would be difficult for a user to type a value By using a record, users can simply select one option from the menu.  
Regarding the other parameters, it is typically easier to know/calculate them.

amaccarini commented on IBPSA/ThermalZones/ISO13790/Examples/FreeFloating.mo:
```diff
@@ -0,0 +1,52 @@
+    Vroo=16*3,
+    f_ms=2.5,
+    redeclare IBPSA.ThermalZones.ISO13790.Data.Light buiMas,
+    gFactor=0.5)
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/Examples/FreeFloatingHVAC.mo:
```diff
@@ -0,0 +1,55 @@
+
+  IBPSA.BoundaryConditions.WeatherData.ReaderTMY3 weaDat(filNam=
+        Modelica.Utilities.Files.loadResource("modelica://IBPSA/Resources/weatherdata/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.mos"))
+    annotation (Placement(transformation(extent={{-80,60},{-60,80}})));
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/Examples/HeatingCooling.mo:
```diff
@@ -0,0 +1,124 @@
+
+
+
+  Modelica.Blocks.Sources.Constant intGains(k=10)
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/Examples/HeatingCooling.mo:
```diff
@@ -0,0 +1,124 @@
+        origin={-34,46},
+        extent={{-6,-6},{6,6}},
+        rotation=0)));
+  Modelica.Blocks.Routing.Multiplex2 multiplex2 annotation (
```
I took the structure of this simple controller from Buildings.ThermalZones.Detailed.Validation.BESTEST.Cases6xx.Case600. Now that I think about it, I am actually not sure why it is there. Maybe @mwetter can answer?

amaccarini commented on IBPSA/ThermalZones/ISO13790/Examples/HeatingCoolingHVAC.mo:
```diff
@@ -0,0 +1,107 @@
+model HeatingCoolingHVAC "Illustrates the use of the 5R1C HVAC thermal zone connected to a ventilation system"
+  extends FreeFloatingHVAC(zonHVAC(nPorts=2));
+
+  Modelica.Thermal.HeatTransfer.Sources.PrescribedHeatFlow preHeaCoo
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/Examples/HeatingCoolingHVAC.mo:
```diff
@@ -0,0 +1,107 @@
+        origin={12,46},
+        extent={{-6,-6},{6,6}},
+        rotation=0)));
+  Fluid.Sources.Boundary_pT bou1(redeclare package Medium = IBPSA.Media.Air,
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/Validation/BESTEST/Case600.mo:
```diff
@@ -0,0 +1,116 @@
+﻿within IBPSA.ThermalZones.ISO13790.Validation.BESTEST;
+model Case600 "Case 600FF, but with dual-setpoint for heating and cooling"
+  extends Case600FF;
+  Modelica.Blocks.Math.Sum sumHeaCoo(nin=2)   annotation (
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/Validation/BESTEST/Case600.mo:
```diff
@@ -0,0 +1,116 @@
+        origin={-24,46},
+        extent={{-6,-6},{6,6}},
+        rotation=0)));
+  Modelica.Blocks.Routing.Multiplex2 multiplex2 annotation (
```
Already answer before

amaccarini commented on IBPSA/ThermalZones/ISO13790/Validation/BESTEST/Case600FF.mo:
```diff
@@ -0,0 +1,46 @@
+within IBPSA.ThermalZones.ISO13790.Validation.BESTEST;
+model Case600FF "Basic test with light-weight construction and free floating temperature"
+  extends Modelica.Icons.Example;
+  Zone5R1C.Zone zon5R1C(
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/Validation/BESTEST/Case900.mo:
```diff
@@ -0,0 +1,116 @@
+﻿within IBPSA.ThermalZones.ISO13790.Validation.BESTEST;
+model Case900
+  extends Case900FF;
+  Modelica.Blocks.Math.Sum sumHeaCoo(nin=2)   annotation (
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/Validation/BESTEST/Case900.mo:
```diff
@@ -0,0 +1,116 @@
+        origin={-24,46},
+        extent={{-6,-6},{6,6}},
+        rotation=0)));
+  Modelica.Blocks.Routing.Multiplex2 multiplex2 annotation (
```
Same as before

amaccarini commented on IBPSA/ThermalZones/ISO13790/Validation/BESTEST/Case900FF.mo:
```diff
@@ -0,0 +1,48 @@
+within IBPSA.ThermalZones.ISO13790.Validation.BESTEST;
+model Case900FF "Test with heavy-weight construction and free floating temperature"
+  extends Modelica.Icons.Example;
+  BoundaryConditions.WeatherData.ReaderTMY3       weaDat(filNam=
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/Zone5R1C/Zone.mo:
```diff
@@ -0,0 +1,333 @@
+  parameter Real nVe(unit="1/h") "Air change rate"
+  annotation (Dialog(group="Ventilation"));
+
+  parameter Modelica.Units.SI.Area[4] Awin "Area of windows"
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/Zone5R1C/Zone.mo:
```diff
@@ -0,0 +1,333 @@
+
+
+
+  Modelica.Thermal.HeatTransfer.Sources.PrescribedTemperature TSup
```
Good suggestion. Changed.

amaccarini commented on IBPSA/ThermalZones/ISO13790/Zone5R1C/Zone.mo:
```diff
@@ -0,0 +1,333 @@
+﻿within IBPSA.ThermalZones.ISO13790.Zone5R1C;
+model Zone "Thermal zone based on 5R1C network"
```
Good point. I think I will leave this aspect for a future "optimization" of the model. Not sure how much a user could change in terms of inputs/outputs, but there are definitely some opportunities, as you mention. I will have a more clear idea after using the model.

amaccarini commented:
Thanks @KremerMartin for your thorough review! I just uploaded all the changes. Hope I answered to all your comments.

KremerMartin commented:
@amaccarini Thanks for revising. I think by checking the IBPSA-library there were added some "minor changes" to many models (e.g. changes in the line breaks or deletion of empty lines). This sometimes happens, when checking the whole library, depending on the configuration of Dymola. I am not sure how it is handled in the IBPSA, but in the AixLib we avoid merging these "accidental" changes into the superior branch.

amaccarini commented:
Oh I see. While pushing the last changes, I could see that many models required an update, but I thought it was because I removed the string about Buildings library, which affected several models. 
I would suggest avoiding to merge these "minor changes". Is it something that I can do? Or that you do as reviewer?

KremerMartin commented:
I think the only way to "redo" the changes is by reverting the last commit and then deleting the String again.

amaccarini commented:
Not very familiar in reverting commits :) Could you please give me some recommendations on how to do that safely?

KremerMartin commented:
@amaccarini I reverted locally at my device and deleted the usage again. (see the last two commits)

mwetter commented on IBPSA/ThermalZones/ISO13790/Zone5R1C/Zone.mo:
```diff
@@ -0,0 +1,327 @@
+    surAzi={surAzi[1],surAzi[2],surAzi[3],
+        surAzi[4],0}) "Solar heat gains of opaque elements"
+    annotation (Placement(transformation(extent={{-100,-100},{-80,-80}})));
+  Modelica.Blocks.Interfaces.RealInput intGai "Internal heat gains"
+    annotation (Placement(transformation(extent={{-180,-140},{-140,-100}})));
```
List all parameters, then all inputs, all output, then the instances, followed by the same order for any protected instances.

amaccarini replied:
Done

mwetter commented on IBPSA/ThermalZones/ISO13790/Zone5R1C/Zone.mo:
```diff
@@ -0,0 +1,327 @@
+network presented in the ISO 13790:2008 Standard. The simplified 5R1C model uses
+five thermal resistances and one thermal capacitance to reproduce the 
+transient thermal behaviour of buildings. The thermal zone is modeled with three 
+temperature nodes, the indoor air temperature <code>Tair</code>, the envelope internal 
+surface temperature <code>Tsur</code> and the building mass temperature Tmas (heat port neglected), and two boundary
+condition nodes, supply air temperature <code>Tsup</code> and the external air temperature 
+<code>Text</code>. The five resistances are related to heat transfer by ventilation <code>Hve</code>, 
```
Use either code as you have now, but then use the right capitalization as in <code>TAir</code>, or use <i>T<sub>air</sub></i>.

amaccarini replied:
Done

mwetter commented on IBPSA/ThermalZones/ISO13790/Zone5R1C/Zone.mo:
```diff
@@ -0,0 +1,327 @@
+</li>
+</ul>
+</html>",
+        info="<html>
+<p>
+Mass data for heavy building
+</p>
+</html>"));
```
Delete duplicate `info` section.

amaccarini replied:
Done

mwetter commented on IBPSA/ThermalZones/ISO13790/Zone5R1C/ZoneHVAC.mo:
```diff
@@ -0,0 +1,73 @@
+    V=VRoo,
+    nPorts=nPorts)       "Air volume"
+              annotation (Placement(transformation(extent={{50,110},{70,130}})));
+  Modelica.Fluid.Vessels.BaseClasses.VesselFluidPorts_b ports_b[nPorts](redeclare
+      each final package
+                    Medium = Medium)
+    "fluid port for adding HVAC system"                         annotation (
```
fix format so it is easier to read. E.g., use
```modelica
 ports_b[nPorts](
  redeclare each final package Medium = Medium)
    "Fluid port for adding HVAC system"
  annotation(...
```

amaccarini replied:
Done

mwetter reviewed (changes requested):
@amaccarini : Thanks for the PR. Please see below for feedback and also check the inline comments. 

Also provide an editable version of the file IBPSA/Resources/Images/ThermalZones/ISO13790/Zone/5R1CNetwork.png

In `GlazedElements` and in `OpaqueElements`, there are four instances (for each direction). Why not using a vector? Is there a benefit of keeping them as numbered instances? 
While for now CPU time will be the same, tool developers are working to handle vectors differently (don't flatten them) which can give faster translation and simulation.

I simplified the BESTEST models to avoid copying and pasting code.
For the BESTEST models it would be nice to indicate the range of the BESTEST publication, similar to what is used in the Buildings library. Then, users would see quickly where we are, and if results change (which frequently do for Buildings if we update Dymola), we can see if we are still in the typical range.

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GainMass.mo:
```diff
@@ -0,0 +1,49 @@
+  Modelica.Blocks.Interfaces.RealInput solGai "Solar gains"
+    annotation (Placement(transformation(extent={{-140,-100},{-100,-60}})));
+  Modelica.Blocks.Interfaces.RealOutput masGai=((AMas/ATot)*(0.5*intGai + solGai)) "Heat gains to mass node"
+    annotation (Placement(transformation(extent={{100,-10},{120,10}})));
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GainSurface.mo:
```diff
@@ -0,0 +1,51 @@
+within IBPSA.ThermalZones.ISO13790.BaseClasses;
+model GainSurface "Surface node heat flow"
+
+  Real AMas "Effective mass area (see Table 12 in standard)";
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GainSurface.mo:
```diff
@@ -0,0 +1,51 @@
+    annotation (Placement(transformation(extent={{-140,-60},{-100,-20}})));
+  Modelica.Blocks.Interfaces.RealInput solGai "Solar gains"
+    annotation (Placement(transformation(extent={{-140,-100},{-100,-60}})));
+  Modelica.Blocks.Interfaces.RealOutput surGai=(1 - AMas/ATot - HWinGai/(9.1*ATot))*(
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GlazedElements.mo:
```diff
@@ -0,0 +1,295 @@
+  parameter Modelica.Units.SI.Angle[4] surAzi
+    "Azimuth angle of surfaces" annotation (Evaluate=true, Dialog(tab="General",
+        group="Window directions"));
+
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/OpaqueElements.mo:
```diff
@@ -0,0 +1,506 @@
+      annotation(Evaluate=true, Dialog(tab = "General", group = "Location"));
+  parameter Modelica.Units.SI.Area[4] AWal "Area of opaque constructions"
+    annotation (Evaluate=true, Dialog(tab="General", group="Construction data"));
+
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/Examples/HeatingCooling.mo:
```diff
@@ -0,0 +1,134 @@
+    "Sum of heating and cooling heat flow rate"
+                                              annotation (
+    Placement(visible = true, transformation(extent={{44,56},{52,64}},      rotation = 0)));
+  Modelica.Thermal.HeatTransfer.Sources.PrescribedHeatFlow preHeaCoo
+    "Prescribed heat flow for heating and cooling"
+    annotation (Placement(visible=true, transformation(extent={{58,54},{70,66}},
+          rotation=0)));
+  Modelica.Blocks.Math.Gain gaiHea(k=1E6) "Gain for heating"
+                                          annotation (Placement(visible=true,
+        transformation(
+        origin={12,72},
+        extent={{-6,-6},{6,6}},
+        rotation=0)));
+  IBPSA.Controls.Continuous.LimPID conHeaPID(
+    Ti=300,
+    k=0.1,
+    reverseActing=true,
+    strict=true) "Controller for heating"
+                 annotation (Placement(visible=true, transformation(
+        origin={-10,72},
+        extent={{-6,-6},{6,6}},
+        rotation=0)));
+  Modelica.Blocks.Sources.Constant TSetHea(k=273.15 + 20)
+    "Set-point for heating"                               annotation (Placement(
+        visible=true, transformation(
+        origin={-34,72},
+        extent={{-6,-6},{6,6}},
+        rotation=0)));
```
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GlazedElements.mo:
```diff
@@ -0,0 +1,295 @@
+  Modelica.Blocks.Math.Gain g3(k=gFac)
+    "g factor of the windows on surface 3"
+    annotation (Placement(transformation(extent={{10,-30},{30,-10}})));
+  Modelica.Blocks.Math.Gain fraWin3(k=1 - winFra)
```
No specific reason, those gains can definitely be merged.

amaccarini replied:
Done

amaccarini commented on IBPSA/ThermalZones/ISO13790/BaseClasses/GlazedElements.mo:
```diff
@@ -0,0 +1,295 @@
+            annotation (Placement(transformation(extent={{-90,100},{-70,120}})));
+  IBPSA.BoundaryConditions.SolarIrradiation.DiffusePerez     HDifTil(
+    til=surTil[1],
+    azi=surAzi[1]) "Diffuse solar irradiation on surface 1"
```
Done

amaccarini commented:
@mwetter, I addressed all comments and the CI tests passed. The major changes are related to the new vectorized approach for adding external surfaces. Now the user can freely choose the number of external surfaces. (before it was a fixed value = 4)

mwetter commented on IBPSA/ThermalZones/ISO13790/Data/Generic.mo:
```diff
@@ -0,0 +1,22 @@
+record Generic "Generic data record building mass"
+   extends Modelica.Icons.Record;
+
+   parameter Real heaC
```
This is not clear, in part because the unit is missing and the documentation does not explain what capacity it is. It is used in 
```modelica
  Modelica.Thermal.HeatTransfer.Components.HeatCapacitor capMas(
      C=buiMas.heaC*AFlo,
      T(displayUnit="degC",
      fixed=true,
      start=293.15)) "Zone thermal capacity"
``` and hence it looks like a capacity per unit area of the floor.

amaccarini replied:
I agree with all changes. The description of `Data.Generic` has been improved. It is a heat capacity per unit area of the floor (J/K m2)

mwetter replied:
Thanks, looks good, but I suggest you change "Heat mass" to "Thermal mass" as the latter is customary.

amaccarini replied:
Not completely sure about what you mean, but I changed the description of `Data.Generic` from "Generic data record building mass" to "Generic data record for building thermal mass". Please let me know if this is what you meant.

mwetter replied:
I updated the data record description in my latest commit to the pull request from issue1492_reducedOrderISO13790_rev_2_mw

mwetter reviewed (changes requested):
@amaccarini : If you agree with my changes, then this is ready to merge, except for the comment on `Data.Generic` which misses the unit and explanation.

I mainly edited the documentation and simplified the example `HeatingCoolingHVAC`. I also reduced the simulation time of these examples (but not of BESTEST) to 10 days. Otherwise, comparing results against other tools becomes very hard due to the frequency of the output.

amaccarini commented:
@mwetter, Thanks a lot for the changes. It seems ready to be merged, but I probably need to merge the parallel branch you created first, right?! Still waiting for the CI tests to be finished.

mwetter commented:
@amaccarini : Yes, best in terms of workflow is to merge my PR to your branch, and then your branch to the master.

amaccarini commented:
@mwetter It looks like there are some issues with the CI tests in your PR. Just let me know if I can help.

amaccarini commented:
@mwetter Thanks a lot for the further revision and updates!

Regarding the value "9.1", I made it as a parameter `hSur`. I think you meant "4.5" ("3.45" is already a parameter). This is a dimensionless coefficient that represents the ratio between internal surfaces and floor area. I made this as a parameter, and move it to protected, as users would not typically change it. 

Other changes are:

- I added a reference paper for the BESTEST validation (it's in the documentation of the package BESTEST)
- Moved the parameter `facMas` to `Data.Generic` as you suggested.
- Changed the dialog box group name of parameters in `Data.Generic` from "Heat mass" to "Thermal mass".

mwetter commented:
- [ ] The icon and diagram view of Zone.mo does not show in OpenModelica. This seems to go back quite a few commits. Commit 703379ced38fbe281d714b4d1cf3a4a22e8ef5a2 which is from September displays it correctly. I will look into what causes the issue.

mwetter commented:
@amaccarini : The model `Zone5R1C` still needs some revisions as it is not entirely clear yet. The issues that should be addressed are
- [ ] There are parameter `nOrientations`, `surTil[:]`, `surAzi[:]`, `AWal[:]` and `AWin[:]`. How are these dimensions relatad? `nOrientations` is used in `win` and `opa`, does this imply that every wall must have a window? This is not clear and should be explained, and added the comment in the parameters.
- [ ] A minimum attribute for `nOrientations` need to be added. Can it be 0, or does not need to be 1 or higher?
- [ ] It looks like `AWal` is only the opaque area of the wall. E.g., if a wall is 4x3 meters, and has a 2x2 window, then a user needs to enter `AWal = {4*3-2*2, ...}`. This should be explained so users don't double count the window area.
- [ ] If there is a constraint on the dimensions of `nOrientations`, `surTil[:]`, `surAzi[:]`, `AWal[:]` and `AWin[:]` then this should be checked with an assertion.

We should also check any potential changes needed so it renders in view of the above OpenModelica issue.

amaccarini commented:
@mwetter Following your suggestions, I made these changes:

- Added assertion to check that parameters `AWal`, `AWin`, `surTil` and `surAzi` have the same dimension of `nOrientations`
- Improved the description of the parameter `AWal` by specifying that the user should only consider opaque parts.
- Added a section in the documentation of `Zone.mo` where I give "Tips for parametrization" regarding the two points above and the possibility to assign the window area equal to 0, if a wall contains only opaque parts. 
- Added minimum attribute to `nOrientations` equal to 1.

amaccarini commented:
Regarding the issue with OpenModelica, if I understand correctly, the problem is related to the fact that the parameters `AWin` etc do not have a specified dimension. Does this mean that the Zone.mo is the only model in the IBPSA library that has unspecified dimension of a parameter?

mwetter commented:
@amaccarini : The model `IBPSA.ThermalZones.ReducedOrder.RC.TwoElements` has the same rendering issue in OMEdit. However, in `Zone5R1C.Zone`, I fixed the issue by setting `nOrientations = 4` as the default. Because we check for the correct entry with the assertions, I think this is fine here.
I also revised the documentation and made sure the parameter names are consistently used in the figure and the mathematical text. From my point of view, this is ready to be merged.

mwetter commented on IBPSA/ThermalZones/ISO13790/package.mo:
```diff
@@ -0,0 +1,34 @@
+    Documentation(info="<html>
+    <p>This package contains models for reduced building physics of thermal zones 
+    based on a thermal network consisting of five resistances and one capacitance. 
+    The models are inspired by the ISO 13790:2008 Standard.</p>
```
@amaccarini : Can you explain what "inspired" means? Are the models based on the standard, or are there differences you want to explain to the user.

amaccarini replied:
@mwetter : The models are based on the ISO13790 Standard. The word "inspired" was not good. I replaced "inspired by" with "based on".

mwetter reviewed (commented):
@amaccarini See inline comment regarding "inspired by the ISO 13790:2008 Standard"

amaccarini commented:
Glad to hear that this is ready to be merged!

@mwetter : Many thanks for the thorough review of these models. I learned a lot about conventions and procedures for developers within the IBPSA library during this review/revision process! 
Thanks again also to @KremerMartin.

@mwetter I have a couple of ideas for the future where I will probably need to add one more example model to show how these models could be parameterized externally in a 3D environment (Blender). But I still need to figure out few things first. I could briefly explain what I have in mind next week at the Modelica IBPSA WG.

mwetter commented:
@amaccarini : Discussing next steps during the coordination meeting next week sounds good. I updated the agenda accordingly.
